erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,824 @@
+        key = copy_struct(tuple_ptr[1], key_size, &hp, &c_p->off_heap);
+        tup = TUPLE2(hp, key, tuple_ptr[2]);
+        hp += 3;
+        res = CONS(hp, tup, res);
+        hp += 2;
+    }
+    erts_free(ERTS_ALC_T_TMP, copy_data);
+    return res;
+}
+
+static BIF_RETTYPE
+persistent_term_info_trap(BIF_ALIST_1)
+{
+    TrapData* trap_data = (TrapData *) BIF_ARG_1;
+    Eterm res;
+
+    ASSERT(is_big(BIF_ARG_1));
+    trap_data = (TrapData *) boxed_val(BIF_ARG_1);
+    res = do_info(BIF_P, trap_data);
+    if (trap_data->remaining > 0) {
+        ASSERT(res == am_ok);
+        BUMP_ALL_REDS(BIF_P);
+        BIF_TRAP1(&persistent_term_info_export, BIF_P, BIF_ARG_1);
+    } else {
+        /*
+         * Decrement ref count (and possibly delete the hash table
+         * and associated literal area).
+         */
+        dec_table_refc(BIF_P, trap_data->table);
+        ASSERT(is_map(res));
+        BIF_RET(res);
+    }
+}
+
+#define DECL_AM(S) Eterm AM_ ## S = am_atom_put(#S, sizeof(#S) - 1)
+#define NUM_INFO_MAP_KEYS 2
+
+static Eterm
+do_info(Process* c_p, TrapData* trap_data)
+{
+    HashTable* hash_table;
+    Uint remaining;
+    Uint idx;
+    Uint max_iter;
+
+    hash_table = trap_data->table;
+    idx = trap_data->idx;
+#if defined(DEBUG) || defined(VALGRIND)
+    max_iter = 50;
+#else
+    max_iter = CONTEXT_REDS;
+#endif
+    remaining = trap_data->remaining < max_iter ? trap_data->remaining : max_iter;
+    trap_data->remaining -= remaining;
+    while (remaining != 0) {
+        if (is_boxed(hash_table->term[idx])) {
+            ErtsLiteralArea* area;
+            area = term_to_area(hash_table->term[idx]);
+            trap_data->memory += sizeof(ErtsLiteralArea) +
+                sizeof(Eterm) * (area->end - area->start - 1);
+            remaining--;
+        }
+        idx++;
+    }
+    trap_data->idx = idx;
+    if (trap_data->remaining > 0) {
+        return am_ok;           /* Dummy return value */
+    } else {
+        Eterm* hp;
+        Eterm res;
+        Eterm tmp;
+        Eterm keys;
+        Uint memory;
+        Uint hsz = 3 + MAP_HEADER_FLATMAP_SZ + NUM_INFO_MAP_KEYS;
+        flatmap_t* mp;
+
+        memory = sizeof(HashTable) + (trap_data->table->allocated-1) * sizeof(Eterm) +
+            trap_data->memory;
+	(void) erts_bld_uint(NULL, &hsz, hash_table->num_entries);
+	(void) erts_bld_uint(NULL, &hsz, memory);
+        hp = HAlloc(c_p, hsz);
+        keys = make_tuple(hp);
+        hp[0] = make_arityval(NUM_INFO_MAP_KEYS);
+        hp[1] = am_count;
+        hp[2] = am_memory;
+        hp += 3;
+        res = make_flatmap(hp);
+        mp = (flatmap_t *) hp;
+        hp += MAP_HEADER_FLATMAP_SZ;
+        mp->thing_word = MAP_HEADER_FLATMAP;
+        mp->size = NUM_INFO_MAP_KEYS;
+        mp->keys = keys;
+	tmp = erts_bld_uint(&hp, NULL, hash_table->num_entries);
+        *hp++ = tmp;
+	tmp = erts_bld_uint(&hp, NULL, memory);
+        *hp++ = tmp;
+        return res;
```

Comment:
This looks wrong. The map value array should follow directly after `mp->keys`, but erts_bld_uint will advance `hp` for bignums of `num_entries` or `memory`.

Reply:
Will fix.